systick.c

```diff
@@ -0,0 +1,100 @@
+/*
+  Not part of Grbl. KeyMe specific.
+
+  SysTick implementation. Global SysTick value is updated.
+  Callbacks can be registed to be called after a certain time.
+
+  SysTick uses Timer1. Timer1 should not be used anywhere else.
+
+*/
+#include "systick.h"
+#include "nuts_bolts.h"
+
+#define MAX_CALLBACKS 32
+
+callback_t systick_callbacks_array[MAX_CALLBACKS];
+uint8_t systick_len;  // Length of callback array 
+
+void systick_service_callbacks()
+{
+  if (!systick_len) {
+    return;
+  }
+  // Assumes the callback array is sorted 
+  // according to lowest callback time
+  if (systick_callbacks_array[0].callback_time <= SysTick) {
```

Comment:
Just summarizing what we discussed this morning:

If we go with the while loop, we have the benefit of having the callbacks being called closer to real-time, but at the risk of having an infinite loop. The risk of having an infinite loop increases when we implement longer callback functions and if we decrease the interval between system ticks. 

For example, if we change sys_tick to 1 us and you have a function that takes 2us to execute and schedules a callback every 1 us, this callback function will be stuck in the while loop indefinitely.

At this point in time, the benefit of using a while loop is not very apparent, neither is the risk of using it, since we have very few callbacks. It is more of a design strategy if we decide to use the callback loop extensively in the future.

This is certainly up for debate, but I tend to err on the safe side and rather call one callback function per iteration of the protocol main loop.